src/debugger/stack-trace.c

```diff
@@ -0,0 +1,100 @@
+/* Copyright (c) 2013-2016 Jeffrey Pfau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include <mgba/internal/debugger/stack-trace.h>
+
+#include <mgba/core/core.h>
+
+#define CHECK_LENGTH() \
+	if (written >= *length) { \
+		*length = written; \
+		return; \
+	}
+
+DEFINE_VECTOR(mStackFrames, struct mStackFrame);
+
+void mStackTraceInit(struct mStackTrace* stack, size_t registersSize) {
+	mStackFramesInit(&stack->stack, 0);
+	stack->registersSize = registersSize;
+}
+
+void mStackTraceDeinit(struct mStackTrace* stack) {
+	mStackTraceClear(stack);
+	mStackFramesDeinit(&stack->stack);
+}
+
+void mStackTraceClear(struct mStackTrace* stack) {
+	ssize_t i = mStackTraceGetDepth(stack) - 1;
+	while (i >= 0) {
+		free(mStackTraceGetFrame(stack, i)->regs);
+		--i;
+	}
+	mStackFramesClear(&stack->stack);
+}
+
+size_t mStackTraceGetDepth(struct mStackTrace* stack) {
+	return mStackFramesSize(&stack->stack);
+}
+
+struct mStackFrame* mStackTracePush(struct mStackTrace* stack, uint32_t pc, uint32_t destAddress, uint32_t sp, void* regs) {
+	struct mStackFrame* frame = mStackFramesAppend(&stack->stack);
+	frame->callAddress = pc;
+	frame->entryAddress = destAddress;
+	frame->frameBaseAddress = sp;
+	frame->regs = malloc(stack->registersSize);
```

Comment:
The `malloc`/`free` during runtime makes me sad, but unfortunately mGBA doesn't have an appropriate utility type for this yet. That can be a later enhancement I suppose.